text/0111-auto-resource-detection.md

```diff
@@ -0,0 +1,190 @@
+configure resource information.
+
+Note there are some existing implementations of this already in the SDKs (see
+[below](#prior-art-and-alternatives)), but nothing currently in the
+specification.
+
+## Explanation
+
+In order to apply auto-detected resource information to all kinds of telemetry,
+a user will need to configure which resource detector(s) they would like to run
+(e.g. AWS EC2 detector).
+
+If multiple detectors are configured, and more than one of these successfully
+detects a resource, the resources will be merged according to the Merge
+interface already defined in the specification, i.e. the earliest matched
+resource's attributes will take precedence. Each detector may be run in
+parallel, but to ensure deterministic results, the resources must be merged in
+the order the detectors were added.
+
+A default implementation of a detector that reads resource data from the
+`OTEL_RESOURCE` environment variable will be included in the SDK. The
+environment variable will contain of a list of key value pairs, and these are
+expected to be represented in a format similar to the [W3C
+Correlation-Context](https://github.com/w3c/correlation-context/blob/master/correlation_context/HTTP_HEADER_FORMAT.md#header-value),
+except that additional semi-colon delimited metadata is not supported, i.e.:
+`key1=value1,key2=value2`. If the user does not specify any resource, this
+detector will be run by default.
+
+Custom resource detectors related to specific environments (e.g. specific cloud
+vendors) must be implemented as packages separate to the core SDK, and users
+will need to import these separately.
+
+## Internal details
+
+As described above, the following will be added to the Resource SDK
+specification:
+
+- An interface for "detectors", to retrieve resource information
+- Specification for a global function to merge resources returned by a set of
+  detectors
+- Details of the "from environment variable" detector implementation as
+  described above
+- Specification that default detection (from environment variable) runs once on
+  startup, and is used by all tracer & meter providers by default if no custom
+  resource is supplied
+
+This is a relatively small proposal so it is easiest to explain the details with
+a code example:
+
+### Usage
+
+The following example in Go creates a tracer and meter provider that uses
+resource information automatically detected from AWS or GCP:
+
+Assumes a dependency has been added on the `otel/api`, `otel/sdk`,
+`otel/awsdetector`, and `otel/gcpdetector` packages.
+
+```go
+resource, _ := sdkresource.Detect(ctx, 5 * time.Second, awsdetector.ec2, gcpdetector.gce)
+tp := sdktrace.NewProvider(sdktrace.WithResource(resource))
+mp := push.New(..., push.WithResource(resource))
+```
+
+### Components
+
+#### Detector
+
+The Detector interface will simply return a Resource:
+
+```go
+type Detector interface {
+    Detect(ctx context.Context) (*Resource, error)
```

Comment:
What is the context used for here?

Reply:
This is just a go specific thing, where context always has to be passed around manually. For most other languages, ignore it. I should probably change these snippets to psuedo-code.

Reply:
The telemetry context (which is indistinguishable from Go's context because it got re-used for that) is very meaningful for OpenTelemetry, so requiring it or not is an important distinction.

Reply:
I suggest we remove it from here, as it seems to be a Go-specific thing ;)

Reply:
> The telemetry context (which is indistinguishable from Go's context because it got re-used for that) is very meaningful for OpenTelemetry, so requiring it or not is an important distinction.

That is a good point and I had not thought about that. I had only included the context in this code snippet for the regular Go usage of context. I don't think there's any requirement to pass the telemetry context into this function, but I'm happy to learn if there is a use case for that.

> I suggest we remove it from here, as it seems to be a Go-specific thing ;)

I removed the code snippets (other than the `Usage` example) & replaced with slightly clearer wording.